src/turing_model.jl

```diff
@@ -0,0 +1,226 @@
+* `"student"`
+* `"bernoulli"`
+* `"poisson"`
+* `"negativebinomial"`
+
+# `priors`
+
+`TuringGLM.jl` comes with state-of-the-art default priors, based on the literature and the
+Stan community.
+By default, `turing_model` will use `DefaultPrior`. But you can specify your own with
+`priors=CustomPrior(predictors, intercept, auxiliary)`. All models take a `predictors` and
+`intercept` priors.
+
+In robust models, e.g. Linear Regression with Student-t likelihood or Count Regression
+with Negative Binomial likelihood, often there is an extra auxiliary parameter that is
+needed to parametrize to model to overcome under- or over-dispersion. If you are specifying
+a custom prior for one of these type of models, then you should also specify a prior for
+the auxiliary parameter.
+
+Non-robust models do not need an auxiliary parameter and you can pass `nothing` as the
+auxiliary argument.
+
+Example for a non-robust model: `@formula(y, ...), data; priors=CustomPrior(Normal(0, 2.5), Normal(10, 5), nothing)`
+
+Example for a robust model: `@formula(y, ...), data; priors=CustomPrior(Normal(0, 2.5), Normal(10, 5), Exponential(1))`
+
+# `standardize`
+
+Whether `true` or `false` to standardize your data to mean 0 and standard deviation 1
+before inference. Some science fields prefer to analyze and report effects in terms of
+standard devations. Also, whenever measurement scales differs, it is often suggested to
+standardize the effects for better comparison. By default, `turing_model` sets `standardize=false`.
+"""
+function turing_model(
+    formula::FormulaTerm, data::D; family="normal", priors=DefaultPrior(), standardize=false
+) where {D}
+
+    # extract y, X and Z
+    y = data_response(formula, data)
+    X = data_fixed_effects(formula, data)
+    Z = data_random_effects(formula, data)
+
+    # μ and σ identities
+    μ_X = 0
+    σ_X = 1
+    μ_y = 0
+    σ_y = 1
+
+    if standardize
+        μ_X, σ_X, X = standardize_predictors(X)
+        μ_y, σ_y, y = standardize_predictors(y)
+    end
+
+    # Random-Effects Conditionals
+    if has_ranef(formula)
+        # Likelihood Conditionals
+        # TODO
+    else
+        # Likelihood Conditionals
+        if family == "normal"
+            if priors isa DefaultPrior
+                custom_prior = CustomPrior(
+                    TDist(3), LocationScale(median(y), mad(y), TDist(3)), nothing
+                )
+            else
+                custom_prior = priors
+            end
+            @model function normal_model(
+                y,
+                X;
+                predictors=size(X, 2),
+                μ_X=μ_X,
+                σ_X=σ_X,
+                prior=custom_prior,
+                residual=1 / std(y),
+            )
+                α ~ prior.intercept
+                β ~ filldist(prior.predictors, predictors)
+                σ ~ Exponential(residual)
+                y ~ MvNormal(α .+ X * β, σ^2 * I)
+                return (; α, β, σ, y)
+            end
```

Comment:
Can this closure be moved into a separate function? It's very difficult to read now whether it's a function call or function definition. E.g.
```
function construct_normal_model(y, X, predictors, ...)
     return @model function normal_model(
           ...
     end
end
```

Reply:
Besides that, it might also be interesting to be able to access the model definition from outside for other purposes. I don't know how realistic that is, but e.g., to allow people to just look at and work with the interiors.

Reply:
Yes! I was thinking in a `turing_code` function to return or print the Turing underlying Turing code.

I think that we can also have that function to be the main one. `turing_model` would just call `turing_code` and eval the string as a macro. Is that possible?

Reply:
Well, ideally that would work in DPPL itself. We have been thinking about adding the original expression to the model struct and use it for printing, which would be pretty minor change but just hasn't been considered to be of high urgency.

However, this will then only work on the instantiated model object, not the result of the `@model` itself, which after all is just a bunch of method definitions of the model evaluator function. 

I wouldn't do eval within the function. If necessary, you could have a top-level dict of names to code and evaluate those on the top level, but I don't really see the advantage.

Reply:
Ok, I will try to work on something during the week. This week is nasty because there are a lot of final committees and also tons of final semester assessments to grade. But it is the final week of the academic semester for me. I will add the `turing_code` suggestion in the comment down below \/